Sources/StreamChat/StateLayer/ChatClient+ChatList.swift

```diff
@@ -0,0 +1,31 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+// MARK: - Factory Methods for Creating Chat Lists
+
+@available(iOS 13.0, *)
+extension ChatClient {
+    /// Creates an instance of ``ChatList`` which represents an array of channels matching to the specified ``ChannelListQuery``.
+    ///
+    /// Matching channels are stored in ``ChatListState.channels``. Use pagination methods in ``ChatList`` for loading more matching channels to the observable state.
+    /// Refer to [querying channels in Stream documentation](https://getstream.io/chat/docs/ios-swift/query_channels/?language=swift) for additional details.
+    ///
+    /// - Note: Only channels that the user can read are returned, therefore, make sure that the query uses a filter that includes such logic. It is recommended to include a members filter which includes the currently logged in user (e.g. `.containMembers(userIds: ["thierry"])`).
+    ///
+    /// - Parameters:
+    ///   - query: The query specifies which channels are part of the list and how channels are sorted.
+    ///   - dynamicFilter: A filter block for filtering by channel's extra data fields or as a manual filter when ``ChatClientConfig.isChannelAutomaticFilteringEnabled`` is false ([read more](https://getstream.io/chat/docs/sdk/ios/client/controllers/channels/)).
+    ///
+    /// - Throws: An error while communicating with the Stream API.
+    /// - Returns: An instance of ``ChatList`` which represents actions and the current state of the list.
+    public func makeChatList(with query: ChannelListQuery, dynamicFilter: ((ChatChannel) -> Bool)? = nil) async throws -> ChatList {
+        let channels = try await channelListUpdater.update(channelListQuery: query)
+        let list = ChatList(channels: channels, query: query, dynamicFilter: dynamicFilter, channelListUpdater: channelListUpdater, client: self)
+        let state = list.state
+        syncRepository.trackChannelListQuery { [weak state] in state?.query }
+        return list
```

Comment:
Not the most happy with this but I feel like it needs to be tackled later. The core idea is that SyncRepository wants to know which queries are active, active are the ones we have a ChatChannelListController instance or a ChatList + ChatListState combo. Whenever ChatListState gets deallocated, the query closure returns nil.

I added a future TODO to rethink this since it is not so nice.

Reply:
While I was looking into this then there seems to be these options:

1. Follow what ChatChannelListController is doing: keep track of allocated instances weakly, when syncing occurs, reset `ChannelListQueryDTO` channels. e.g. `queryDTO.channels.removeAll()`. Con: is ChatClient+SyncRepository needs to track instances. Pro: less updating, no need to figure out how to store `ChannelListQuery` which is required for resetting/reloading.
2. Remove tracking of active controllers/queries  from `SyncRepository` and instead, `SyncRepository` could fetch all the `ChannelListQueryDTO`s and reset `channels` properties. Pro: no coupling Con: controller/ChatList need to figure out when resetting happens and then triggering reload themselves. The challenge is that we need `ChannelListQuery` for loading and only controller and ChatList knows that.
3. SyncRepository does not reset any channels for queries and instead sends some event which ChatChannelListController and ChatList listen to. When this happens, they call `resetChannelsQuery` themselves. Pro: no weakly tracking Con: reset requires `watchedAndSynchedChannelIds`, `synchedChannelIds` arguments which are managed by SyncRepository which makes it more challenging.

I am happy to try 2 or 3 but I feel like it should then be a separate PR for the feature branch.